File: src_java/NeuralNet.java
package fnn;

import java.util.Arrays;

/**
 * 1-hidden neural network for _classification_
 *
 */

public class NeuralNet {
  private float[][] trainingData;
  private float[][] testingData;
  private int nbInstances;
  private int nbFeatures;
  private int nbClasses;
  private int batchSize;
  
  private float[][] X_train, Y_train, X_test, Y_test, W1, W2, b1, b2;
  private float eta = 0.01f;

  /**
   * Class constructor
   * @param data      data read from input file
   * @param batchSize number of instances in each batch for training
   * @param K         number of output class (e.g. Iris dataset => 3)
   */
  public NeuralNet(float[][] data, int batchSize, int K /*nb classes*/) {
    nbInstances = data.length;
    nbFeatures  = data[0].length;
    this.batchSize = batchSize;
    this.nbClasses = K;
    
    int trainingSize = (int) (nbInstances*0.75);
    int testingSize  = nbInstances-trainingSize;
    trainingData = new float[trainingSize][nbFeatures];
    testingData  = new float[testingSize][nbFeatures];
    
    // Copy data into training & testing set
    for (int i = 0; i < trainingSize; i++)
      for (int j = 0; j < data[0].length; j++)
        trainingData[i][j] = data[i][j];
    
    for (int i = 0; i < testingSize; i++)
      for (int j = 0; j < data[0].length; j++)
        testingData[i][j] = data[i+trainingSize][j];
    
    
    X_train = new float[this.nbFeatures-1][this.batchSize];  // -1 because the last column is the label
    Y_train = new float[this.nbClasses][this.batchSize];
    
    X_test  = new float[this.nbFeatures-1][this.batchSize];
    Y_test  = new float[this.nbClasses][this.batchSize];
    
    W1 = new float[3][4]; NNLib.initMatrix(W1);
    b1 = new float[3][1]; for (int i = 0; i < b1.length; i++) Arrays.fill(b1[i], 0.f);
    
    W2 = new float[3][3]; NNLib.initMatrix(W2);
    b2 = new float[3][1]; for (int i = 0; i < b2.length; i++) Arrays.fill(b2[i], 0.f);
  }
  
  @SuppressWarnings("unused")
  private void printData(float[][] data){
    for (int i = 0; i < data.length;i++){
      for (int j = 0; j < data[0].length; j++)
        System.out.print(data[i][j] + " ");
      System.out.println();
    }
  }
  
  /**
   * Create one-hots vector for classification
   * @param k target      class to set to 1
   * @param batchInstance current data point in Y (a column index)
   * @param Y             matrix of one-hots
   */
  private void createOneHot(int k, int indexInBatch, float[][] Y){
    if (k > Y.length) throw new RuntimeException("There is only " + Y.length + " possible classes");
    /* TODO Part 3, Q.1 */
    int i;
    for(i=0;i<Y.length;i++){
    	if(i==k){
    		Y[i][indexInBatch]=1;
    	}
    	else{
    		Y[i][indexInBatch]=0;
    	}
    }
  }
  
  /**
   * Load one instance of data
   * Separate features and labels #dataIndex in X and Y
   * @param dataSet   training or testing
   * @param X         matrix for inputs
   * @param Y         matrix for outputs
   * @param dataIndex index in dataSet of data to copy
   */
  private void loadAttributesAndLabels(float[][] dataSet, float[][] X, float[][] Y, int dataIndex, int batchSize){
    // 1. Load data attributes in X
    int i = 0; // index to load the features (data[ex][i] => feature #i of instance #ex)
    int indexInBatch = dataIndex%batchSize;
    /* TODO Part 3, Q.2 */
    for(i=0;i<dataSet[0].length-1;i++){
    	X[indexInBatch][i] = dataSet[indexInBatch][i];
    }
    
    // 2. Load data labels in Y (create a one-hot for class prediction)
    createOneHot((int) dataSet[dataIndex][i], indexInBatch, Y);
  }
  
  /**
   * Shuffle the set of training data
   * Fisher–Yates shuffle :
   * https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm
   * Note : NNLib.rnd.nextInt(max) returns an integer in [0;max[
   */
  private void shuffleTrainingData() {
    /* TODO Part 4, Q.1 */
	  int i,j,k;
	  float tmp;
	  for(i=trainingData[0].length;i>=1;i--) {
		  j = NNLib.rnd.nextInt(i+1);
		  for(k=0;k<nbFeatures;k++){
			  tmp = trainingData[i][k];
			  trainingData[i][k] = trainingData[j][k];
			  trainingData[j][k] = tmp;
		  }
	  }
  }
  
  /**
   * Feed the testing set to the model and measure error / accuracy
   * @returns cost on testing set
   */
  public float testPrediction() {
    int currentDataIndex = 0;
    float cost = 0.f;
    int countPredictions = 0;
    
    /* TODO Part 4, Q.3 */
    int i;
    for(i=0;i<testingData[0].length;i++) {
    	loadAttributesAndLabels(testingData,X_test,Y_test,i,batchSize);
    }
    float[][] Z1 = NNLib.addVec(NNLib.mult(W1,X_test),b1); 
	float[][] A1 = NNLib.tanh(Z1);
	float[][] Z2 = NNLib.addVec(NNLib.mult(W2, A1), b2);
	float[][] A2 = NNLib.softmax(Z2);
	float trainingError = NNLib.crossEntropy(A2, Y_test);
   
	//System.out.println("%d",(int)trainingError);
    cost /= testingData.length;
    float accuracy = 100.f*countPredictions/testingData.length;
    System.out.println("  CE cost on test data: "  + cost);
    System.out.println("  Accuracy on test data: " + accuracy);
    return cost; //accuracy;
  }
  
  /**
   * Perform 1 epoch of training
   *  1/ load a minibatch of data into X_train and Y_train
   *  2/ forward pass
   *  3/ compute the gradient of the loss
   *  4/ update the parameters
   */
  private float trainingEpoch(){
    int seenTrainingData = 0;
    @SuppressWarnings("unused")
    float trainingError  = 0.f;
    float testingError   = 0.f;
    
    shuffleTrainingData(); // shuffle the data before training
    
    /* TODO Part 4, Q.2 */
    int i;
    while(seenTrainingData<trainingData.length){
    	for(i=0;i<batchSize;i++){
    		loadAttributesAndLabels(trainingData,X_train,Y_train,i,batchSize);
    	}
    	/* Propagation avant */
    	float[][] Z1 = NNLib.addVec(NNLib.mult(W1,X_train),b1); 
    	float[][] A1 = NNLib.tanh(Z1);
    	float[][] Z2 = NNLib.addVec(NNLib.mult(W2, A1), b2);
    	float[][] A2 = NNLib.softmax(Z2);
    	trainingError = NNLib.crossEntropy(A2, Y_train);
    	
    	/* Rétropropagation */
    	float[][] delta2 = NNLib.subtract(A2, Y_train);
    	float[][] dW2 = NNLib.mult(delta2, NNLib.transpose(A1));
    	float[][] db2 = delta2;
    	
    	float[][] delta1 = NNLib.hadamard(NNLib.mult(W2,delta2), NNLib.tanhDeriv(Z1));
    	float[][] dW1 = NNLib.mult(delta1, NNLib.transpose(X_train));
    	float[][] db1 = delta1;
    	
    	/* Mise à jour des paramètres */
    	W2 = NNLib.subtract(W2, NNLib.mult(dW2, eta));
    	b2 = NNLib.subtract(b2,NNLib.mult(db2,eta));
    	
    	W1 = NNLib.subtract(W1, NNLib.mult(dW1, eta));
    	b1 = NNLib.subtract(b1,NNLib.mult(db1,eta));
    	
    	seenTrainingData+=batchSize;
    }
    
    return testPrediction();
  }
  
/**
   * Train the neural network
   * @param nbEpoch number of epoch to run
   */
  public void train(int nbEpoch){
    String trainingProgress = "";
    for (int e = 0; e < nbEpoch; e++){
      System.out.println(" [ Epoch " +e+ "]");
      trainingProgress += e + "," + trainingEpoch() + "\n";
    }
    // Export the error per epoch in output file
    DataLib.exportDataToCSV("training.out", trainingProgress);
  }

}

File: src/NeuralNet.py
#!/usr/bin/python

import numpy as np
import random
from DataLib import DataLib
from NNLib import NNLib

class NeuralNet:
    eta = 0.01

    def __init__(self,data,batchsize,K, hiddenlayers, hiddenlayersizes):
        self.nbInstances = data.shape[0]
        self.nbFeatures = data.shape[1]-1
        self.batchSize = batchsize
        self.nbClasses = K
        self.trainingData = data
        self.hiddenLayers = hiddenlayers
        self.hiddenLayersSizes = hiddenlayersizes
        self.trainingSize = (int)(0.74*self.nbInstances)
        self.testingSize = self.nbInstances - self.trainingSize
        self.trainingData = data[0:self.trainingSize][:]

        self.testingData = data[self.trainingSize:self.nbInstances][:]
        # self.testingData = data
        for i in range(self.hiddenLayers+1):
            if i==0:
                self.W = [np.random.rand(self.nbFeatures,self.hiddenLayersSizes[0])]
                self.b = [np.random.rand(self.batchSize,self.hiddenLayersSizes[0])]
            elif i==self.hiddenLayers:
                self.W = self.W + [np.random.rand(self.hiddenLayersSizes[i-1],self.nbClasses)]
                self.b = self.b + [np.random.rand(self.batchSize,self.nbClasses)]
            else:
                self.W = self.W + [np.random.rand(self.hiddenLayersSizes[i-1],self.hiddenLayersSizes[i])]
                self.b = self.b + [np.random.rand(self.batchSize,self.hiddenLayersSizes[i])]
        
        # to verify weight matrix sizes and bias matrix sizes
        # for i in range(self.hiddenLayers+1):
        #     print(self.W[i].shape)
        
        # for i in range(self.hiddenLayers+1):
        #     print(self.b[i].shape)

        self.X_train = np.empty([self.batchSize,self.nbFeatures])
        self.Y_train = np.empty([self.batchSize,self.nbClasses])
        self.X_test = np.empty([self.batchSize,self.nbFeatures])
        self.Y_test = np.empty([self.batchSize,self.nbClasses])

        self.result = 0
        self.falsePos = 0
        self.falseNeg = 0
        self.truePos = 0
        self.trueNeg = 0

        self.errors = []

    #done
    # if label = 0 then one-hot = (1,0)
    # if label = 1 then one-hot = (0,1)
    def createOneHot(self,k,indexInBatch,Y):
        if(k>Y.shape[1]-1): # if k is higher than the number of columns
            print("There is only " + str(Y.shape[1]) + " possible classes!")
        for i in range(Y.shape[1]):
            if i==k:
                Y[indexInBatch][i] = 1
            else:
                Y[indexInBatch][i] = 0
    
    #done
    def shuffleTainingData(self):
        np.random.shuffle(self.trainingData)

    # done
    def loadAttributesAndLabels(self,dataSet,X,Y,dataIndex,batchSize):
        i = 0
        indexInBatch = dataIndex%batchSize
        for i in range(dataSet.shape[1]-1): # -1 because we do not want to take the label with the features
            X[indexInBatch][i] = dataSet[dataIndex][i]
        i += 1
        self.createOneHot((int)(dataSet[dataIndex][i]),indexInBatch,Y)
    
    # not finished
    def testPrediction(self):
        seenTestingData = 0
        offset = 0
        bSize = 1

        while seenTestingData<self.testingData.shape[0]:
            for i in range(bSize):
                self.loadAttributesAndLabels(self.testingData,self.X_test,self.Y_test,offset+i,bSize)           
            offset += bSize

            Z = []
            A = []


            #Forward propagation
            # testing if the result matches
            # (1,0) = not sick, (0,1) = sick

            for i in range(self.hiddenLayers+1):
                if i == 0:
                    Z.append((self.X_test @ self.W[0]) + self.b[0])
                else:
                    Z.append((A[i-1] @ self.W[i]) + self.b[i])
                
                if i == self.hiddenLayers:
                    A.append(NNLib.softMax(Z[i]))
                else:
                    A.append(NNLib.tanh(Z[i]))


            # testing if the result matches
            if self.Y_test[0][0]==1 and A[self.hiddenLayers][0][0]>=0.5:
                self.result += 1
                self.trueNeg += 1
            elif self.Y_test[0][1]==1  and A[self.hiddenLayers][0][1]>=0.5 :
                self.result += 1
                self.truePos += 1
            elif self.Y_test[0][0]==0 and A[self.hiddenLayers][0][0]>=0.5 :
                self.falseNeg += 1
            else:
                self.falsePos += 1
            
            seenTestingData += bSize

        print("TOTAL OF CORRECT PREDICTIONS: " + str(self.result))
        print("TOTAL OF PREDICTIONS: " + str(self.testingData.shape[0]))
        print("PERCENTAGE OF CORRECT PREDICTIONS: " + str((self.result/self.testingData.shape[0])*100) + " %")
        print("TRUE POSITIVES : " + str(self.truePos) + ", Percentage: "+str(self.truePos*100/self.testingData.shape[0])+ " %")
        print("TRUE NEGATIVES : " + str(self.trueNeg) + ", Percentage: "+str(self.trueNeg*100/self.testingData.shape[0])+ " %")
        print("FALSE POSITIVES : " + str(self.falsePos) + ", Percentage: "+str(self.falsePos*100/self.testingData.shape[0])+ " %")
        print("FALSE NEGATIVES : " + str(self.falseNeg) + ", Percentage: "+str(self.falseNeg*100/self.testingData.shape[0])+ " %")
        if self.truePos + self.trueNeg + self.falsePos + self.falseNeg == self.testingData.shape[0]:
            print("NUMBERS MATCHES")

    # done
    def trainingEpoch(self):
        seenTrainingData = 0
        trainingError = 0.0
        testingError = 0.0
        offset = 0
        self.error = []
        
        self.shuffleTainingData()

        while seenTrainingData<self.trainingData.shape[0]:
            for i in range(self.batchSize):
                self.loadAttributesAndLabels(self.trainingData,self.X_train,self.Y_train,offset+i,self.batchSize)
            offset += self.batchSize
            seenTrainingData += self.batchSize

            Z = []
            A = []
            delta = []
            dW = []
            db = []

            #Forward propagation
            for i in range(self.hiddenLayers+1):
                if i == 0: # if first layer
                    Z.append((self.X_train @ self.W[i]) + self.b[i])
                else:
                    Z.append((A[i-1] @ self.W[i]) + self.b[i])
                
                if i != self.hiddenLayers: # if hidden layer
                    A.append(NNLib.tanh(Z[i]))
                else: # if last layer
                    A.append(NNLib.softMax(Z[i]))
                    
                    #Error computing
                    trainingError = NNLib.crossEntropy(A[i],self.Y_train)
                    self.errors = np.append(self.errors,trainingError)  
            
            #Retropropagation of error
            for i in range(self.hiddenLayers+1):
                if i == 0: # if last layer
                    delta.append(A[self.hiddenLayers] - self.Y_train)
                else: # if hidden layer
                    delta.append(delta[i-1] @ np.transpose(self.W[self.hiddenLayers - i + 1]) * NNLib.tanhDeriv(Z[self.hiddenLayers - i]))
                
                db.append(delta[i])

                
                if i != self.hiddenLayers:
                    dW.append(np.transpose(A[self.hiddenLayers - i - 1]) @ delta[i])
                else:
                    dW.append(np.transpose(self.X_train) @ delta[i])
            


            
            #Parameters update
            for i in range(self.hiddenLayers+1):
                self.W[i] = self.W[i] - self.eta * dW[self.hiddenLayers - i]
                self.b[i] = self.b[i] - self.eta * db[self.hiddenLayers - i]

        return np.mean(self.errors)


    # done
    def train(self,nbEpoch):
        trainingProgress = ""
        for e in range(nbEpoch):
            if e % 100 == 0:
                print(" Epoch "+str(e))
            error = self.trainingEpoch()
            # writes error with epoch number into new CSV to make gnuplot graph
            DataLib.writeToCSV(error,e)
        print("Testing the model with the testingData\n")
        self.testPrediction()
